From 9e01389eaa45eb388a0f43e01b77f5b3ae86d9de Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:18:35 -0400
Subject: [net] netfilter: nf_tables: fix wrong transaction ordering in set elements

Message-id: <20141021161835.22888.44196.stgit@localhost.localdomain>
Patchwork-id: 97961
O-Subject: [RHEL7 net PATCH 063/124] netfilter: nf_tables: fix wrong transaction ordering in set elements
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: 46bbafceb201df635b7c0a9d7a0e526cb2f8cb75

 commit 46bbafceb201df635b7c0a9d7a0e526cb2f8cb75
 Author: Pablo Neira Ayuso <pablo@netfilter.org>
 Date:   Thu May 22 12:36:03 2014 +0200

    netfilter: nf_tables: fix wrong transaction ordering in set elements

    The transaction needs to be placed at the end of the commit list,
    otherwise event notifications are reordered and we may crash when
    releasing object via call_rcu.

    This problem was introduced in 60319eb ("netfilter: nf_tables: use new
    transaction infrastructure to handle elements").

    Reported-by: Arturo Borrero Gonzalez <arturo.borrero.glez@gmail.com>
    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/netfilter/nf_tables_api.c b/net/netfilter/nf_tables_api.c
index 0478847..9365531 100644
--- a/net/netfilter/nf_tables_api.c
+++ b/net/netfilter/nf_tables_api.c
@@ -3077,7 +3077,7 @@ static int nft_add_set_elem(struct nft_ctx *ctx, struct nft_set *set,
   goto err4;
 
  nft_trans_elem(trans) = elem;
- list_add(&trans->list, &ctx->net->nft.commit_list);
+ list_add_tail(&trans->list, &ctx->net->nft.commit_list);
  return 0;
 
 err4:
@@ -3161,7 +3161,7 @@ static int nft_del_setelem(struct nft_ctx *ctx, struct nft_set *set,
   goto err2;
 
  nft_trans_elem(trans) = elem;
- list_add(&trans->list, &ctx->net->nft.commit_list);
+ list_add_tail(&trans->list, &ctx->net->nft.commit_list);
 
  nft_data_uninit(&elem.key, NFT_DATA_VALUE);
  if (set->flags & NFT_SET_MAP)
-- 
1.7.1